skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,412 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+"""_max_tree.pyx - building a max-tree from an image.
+
+This is an implementation of the max-tree, which is a morphological
+representation of the image. Many morphological operators can be built
+from this representation, namely attribute openings and closings.
+
+This file also contains implementations of max-tree based filters and
+functions to characterize the tree components. 
+"""
+
+import numpy as np
+
+from libc.stdlib cimport free, malloc, realloc
+from scipy.constants import point
+from cProfile import label
+
+cimport numpy as np
+cimport cython
+
+ctypedef np.float64_t DTYPE_FLOAT64_t
+ctypedef np.int32_t DTYPE_INT32_t
+ctypedef np.uint32_t DTYPE_UINT32_t
+ctypedef np.uint64_t DTYPE_UINT64_t
+ctypedef np.int64_t DTYPE_INT64_t
+ctypedef np.uint8_t DTYPE_BOOL_t
+ctypedef np.uint8_t DTYPE_UINT8_t
+ctypedef np.int8_t DTYPE_INT8_t
+
+ctypedef fused dtype_t:
+    np.uint8_t
+    np.uint16_t
+    np.uint32_t
+    np.uint64_t
+    np.int8_t
+    np.int16_t
+    np.int32_t
+    np.int64_t
+    np.float32_t
+    np.float64_t
+
+
+cdef DTYPE_INT64_t find_root(DTYPE_INT64_t[::1] parent, DTYPE_INT64_t index):
+    """recursive function to get the root of the current tree. Importantly, the
+    function changes the tree (path compression), that reduces the complexity
+    from O(n*n) to O(n*log(n)).
+    """
+    if parent[index] != index:
+        parent[index] = find_root(parent, parent[index])
+    return parent[index]
```

Comment:
I'm not sure whether this
```python
while parent[index] != index:
    parent[index] = parent[parent[index]]
return parent[index]
```
will do the same thing. But if so, I think this would be more space efficient (stack wise) than recursion and allow the function to be inlined.

Reply:
Thanks for the comment. That's very interesting. 

First, no it is not equivalent, but I get the idea. Actually, the code does not terminate, as `index` is not updated. But the following code works:

```python
while parent[index] != parent[parent[index]]:
    parent[index] = parent[parent[index]]
return parent[index]
```

This code leads to the same max-tree, but it is still different in that there is no path compression: in the recursive formulation, all entries of parent along the path between `index` and `root` will point to `root` at the end. In the code above, the `parent` array is only updated at `index`. Algorithmically, the code you proposed should therefore be less performant. 

_But_: you are completely right that there is something to be gained by having a non-recursive function that can be inlined. I tested this, and on `skimage.data.camera()` for instance I gain 20 ms (7 %) with your version. So it seems that inlining more than compensates what is lost by path compression. However, I have no clear view on what happens on bigger images.

Reply:
So the ideal solution would be stack efficient *and* do compression. Right now I'm asking myself if tail-recursion could be applied to this problem. My understanding is that the compiler is be able to optimize tail-recursion away. However [Wikipedia suggests that tail-recursion is related to while-constructs](https://en.wikipedia.org/wiki/Tail_call#Relation_to_while_construct) so I'm not sure if we could gain anything this way.
If the non-recursive solution is inefficient for larger images it might be worth looking into.

Reply:
Shall we keep the two implementations as a reminder for future optimization? Or shall we keep only the faster one?

Reply:
I'd be in favor of adding a note section to the function's docstring describing the alternate approach.

Reply:
OK perfect. I just modify the docstring.